### ISSUE TITLE
KeyPackage lifetime handling

### ISSUE THREAD

**Issue by tomleavy** [mlswg/mls-protocol#538]
The RFC currently says the following:

A client MUST NOT use the data in a KeyPackage for any processing before the not_before date, or after the not_after date.

Clarification would be helpful here because this could be interpreted in different ways and I imagine this would be a point of contention when trying to gain compatibility between implementations. 

Examples: 

* A key package expires after it has been placed into a leaf
* A key package has an expiration in 1 month, a commit is made which places it into a leaf, and a member of the group receives that commit 1 month later due to them being offline

**Comment by kkohbrok** [comment]
Good point and good examples as well. Unfortunately, we're using KeyPackages and the key material in it in two roles: a) as pre-published key material that can be used by others to add new members to a group and b) as as the key material representing a member in a group. I always thought of key package expiration as relevant for role a), where when adding a member, their key material has to be reasonably fresh, since for pre-published keys, the lifetime limit is the only way a client can know that they have achieved PCS at some point (without trusting the server and assuming there is no trusted and reliable revocation mechanism). For role b) it's more of a tracker of when a given party has last made an update.

I would imagine example 1 is more of a policy question. If a key package expires while it's already in a group (role b), the policy could be to remove that member (and maybe even to prohibit any group activity until this was done) or to not accept any messages from that member other than an update, etc.

Due to asynchronicity, I don't think we will be able to reliably solve example 2. We could have the committer (and maybe even the DS) add a time-stamp, so the receiver can more easily distinguish between themselves receiving an old commit and the committer trying to do something fishy, but there we're also relying on clocks 🤷🏻.

Depending on what we want for the spec, it might be worth to have different structs for the two roles. That would also enable us to get slightly better FS guarantees as suggested by @psyoptix a while ago by using different key pairs for group participation and Welcome encryption.

**Comment by tomleavy** [comment]
That makes sense to me @kkohbrok , I touched on this a little bit in #537 in that it `ParentHash` ext seems like it deserves to be a field rather than an extension. If we made that change you have a scenario where `KeyPackage` always does not have a `ParentHash` ext when used to add someone, and it always has a `ParentHash` ext when it is situated in a tree, so different types make sense.

I would say though we would have to be careful to not break `ParentHash` functionality or signatures since I believe there is a step when you join a group where you verify all the key package signatures in the tree

**Comment by tomleavy** [comment]
@kkohbrok the one concern I had about not checking the lifetime upon receiving an add was just that I have no way to verify the sender of the add actually did a reasonable check. I don't see a way to resolve this problem though tbh

**Comment by kkohbrok** [comment]
> @kkohbrok the one concern I had about not checking the lifetime upon receiving an add was just that I have no way to verify the sender of the add actually did a reasonable check. I don't see a way to resolve this problem though tbh

Yes, that one's tricky and I don't think there's a way to solve it properly. However, if you mandate by policy that the sender includes a time stamp (e.g. via a `time_of_add` extension or something similar) and you don't assume the sender is actively malicious or their clock is totally out-of-whack, then you can compare that with the expiration date of the KeyPackage included in the Add to get some idea of whether the KeyPackage was valid when the commit was done. This reminds me that committers should probably check all Add proposals for expired KeyPackages, since the proposal could have been sent well in advance of the actual commit. In any case, it would probably make sense to cover this explicitly either in the spec or in the architecture doc.

**Comment by tomleavy** [comment]
Closing in favor of #539 . The separation there will make it obvious when Lifetime matters
